### PR TITLE
Add support for Packages API

### LINKED ISSUE #2075: Does not to be any support for the Packages API
I do not see any support for the [Packages API](https://docs.github.com/en/rest/reference/packages). I don't know if I miss it somewhere though.

If it does not exist then I would be happy to work on it.

### PR DESCRIPTION
**Pull request by Cyb3r-Jak3**
Adds support for [Packages API](https://docs.github.com/en/rest/reference/packages) through expansion of Users and Organization Services.

Would resolve #2075

### CODE REVIEW

File: github/packages.go [hunk 1]
```diff
@@ -99,3 +106,40 @@ type PackageRegistry struct {
...
+	Container   *PackageContainerMetadata `json:"container,omitempty"`
+}
+
+func (r PackageMetadata) String() string {
+	return Stringify(r)
+}
+
+// PackageContainerMetadata represents container metadata for docker container packages
+type PackageContainerMetadata struct {
+	Tags []*string `json:"tags,omitempty"`
```
Comment (gmlewis) [inline github/packages.go:140]:
```suggestion
	Tags []string `json:"tags,omitempty"`
```

File: github/users_packages.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	if user != "" {
+		u = fmt.Sprintf("user/%v/packages", user)
+	} else {
+		u = "user/packages"
+	}
+	u, err := addOptions(u, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:29]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-for-a-user
+func (s *UsersService) GetPackage(ctx context.Context, user, packageType, packageName string) (*Package, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v", user, packageType, packageName)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v", packageType, packageName)
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:53]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	var pack *Package
```
Comment (gmlewis) [inline github/users_packages.go:57]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	var packages []*Package
+	resp, err := s.client.Do(ctx, req, &packages)
+	if err != nil {
+		return nil, resp, err
+	}
+	return packages, resp, nil
+}
+
+// Get a package by name for a user. Passing the empty string will
```
Comment (gmlewis) [inline github/users_packages.go:41]: Please replace `Passing the empty string will`
with: `Passing the empty string for "user" will`
in this place and all places throughout this PR for consistency
since there are many places where multiple arguments are passed.

File: github/users_packages.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-for-a-user
+func (s *UsersService) DeletePackage(ctx context.Context, user, packageType, packageName string) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v", user, packageType, packageName)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v", packageType, packageName)
+	}
+	req, err := s.client.NewRequest("DELETE", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:77]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/users_packages.go:81]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
```
Comment (gmlewis) [inline github/users_packages.go:82]: Please delete this blank line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-for-a-user
+func (s *UsersService) RestorePackage(ctx context.Context, user, packageType, packageName string) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/restore", user, packageType, packageName)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v/restore", packageType, packageName)
+	}
+	req, err := s.client.NewRequest("POST", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:97]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/users_packages.go:101]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
```
Comment (gmlewis) [inline github/users_packages.go:102]: Please delete this blank line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-all-package-versions-for-a-package-owned-by-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/users#delete-an-email-address-for-the-authenticated-user
+func (s *UsersService) PackageGetAllVersions(ctx context.Context, user, packageType, packageName string) ([]*PackageVersion, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions", user, packageType, packageName)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v/versions", packageType, packageName)
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:117]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	var versions []*PackageVersion
```
Comment (gmlewis) [inline github/users_packages.go:121]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+	resp, err := s.client.Do(ctx, req, &versions)
+	if err != nil {
+		return nil, resp, err
+	}
+	return versions, resp, nil
+
```
Comment (gmlewis) [inline github/users_packages.go:127]: Please delete this blank line.

File: github/users_packages.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// Get a specific version of a package in for a user. Passing the empty string will
```
Comment (gmlewis) [inline github/users_packages.go:130]:
```suggestion
// Get a specific version of a package for a user. Passing the empty string for "user" will
```

File: github/users_packages.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-version-for-a-user
+func (s *UsersService) PackageGetVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*PackageVersion, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v", user, packageType, packageName, packageVersionID)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v/versions/%v", packageType, packageName, packageVersionID)
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:142]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	var version *PackageVersion
```
Comment (gmlewis) [inline github/users_packages.go:146]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+	resp, err := s.client.Do(ctx, req, &version)
+	if err != nil {
+		return nil, resp, err
+	}
+	return version, resp, nil
+
```
Comment (gmlewis) [inline github/users_packages.go:152]: Please delete this blank line.

File: github/users_packages.go [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-package-version-for-a-user
+func (s *UsersService) PackageDeleteVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v", user, packageType, packageName, packageVersionID)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v/versions/%v", packageType, packageName, packageVersionID)
+	}
+	req, err := s.client.NewRequest("DELETE", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:167]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/users_packages.go:171]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+
```
Comment (gmlewis) [inline github/users_packages.go:172]: Please delete this blank line.

File: github/users_packages.go [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-package-version-for-a-user
+func (s *UsersService) PackageRestoreVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v/restore", user, packageType, packageName, packageVersionID)
+	} else {
+		u = fmt.Sprintf("user/packages/%v/%v/versions/%v/restore", packageType, packageName, packageVersionID)
+	}
+	req, err := s.client.NewRequest("POST", u, nil)
```
Comment (gmlewis) [inline github/users_packages.go:187]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/users_packages.go:191]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/users_packages.go [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+
```
Comment (gmlewis) [inline github/users_packages.go:192]: Please delete this blank line.

File: github/orgs_packages.go [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+	u := fmt.Sprintf("orgs/%v/packages", org)
+	u, err := addOptions(u, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	var packages []*Package
```
Comment (gmlewis) [inline github/orgs_packages.go:26]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+// license that can be found in the LICENSE file.
+
+package github
+
+import (
+	"context"
+	"fmt"
+)
+
+// List the packages for an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:13]:
```suggestion
// List the packages for an organization.
```

File: github/orgs_packages.go [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+// Get a package by name from an organization
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-for-an-organization
+func (s *OrganizationsService) GetPackage(ctx context.Context, org, packageType, packageName string) (*Package, *Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v", org, packageType, packageName)
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	var pack *Package
```
Comment (gmlewis) [inline github/orgs_packages.go:43]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+	resp, err := s.client.Do(ctx, req, &pack)
+	if err != nil {
+		return nil, resp, err
+	}
+	return pack, resp, nil
+}
+
+// Delete a package from an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:51]:
```suggestion
// Delete a package from an organization.
```

File: github/orgs_packages.go [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+// Restore a package to an organization
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-for-an-organization
+func (s *OrganizationsService) RestorePackage(ctx context.Context, org, packageType, packageName string) (*Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v/restore", org, packageType, packageName)
+	req, err := s.client.NewRequest("POST", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/orgs_packages.go:73]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+
+}
+
+// Get all versions of a package in an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:77]:
```suggestion
// Get all versions of a package in an organization.
```

File: github/orgs_packages.go [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-all-package-versions-for-a-package-owned-by-an-organization
+func (s *OrganizationsService) PackageGetAllVersions(ctx context.Context, org, packageType, packageName string, opts *PackageListOptions) ([]*PackageVersion, *Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v/versions", org, packageType, packageName)
+	u, err := addOptions(u, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+	req, err := s.client.NewRequest("GET", u, nil)
```
Comment (gmlewis) [inline github/orgs_packages.go:86]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	var versions []*PackageVersion
```
Comment (gmlewis) [inline github/orgs_packages.go:90]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+	resp, err := s.client.Do(ctx, req, &versions)
+	if err != nil {
+		return nil, resp, err
+	}
+	return versions, resp, nil
+
```
Comment (gmlewis) [inline github/orgs_packages.go:96]: Please delete this blank line.

File: github/orgs_packages.go [hunk 34]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Get a specific version of a package in an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:99]:
```suggestion
// Get a specific version of a package in an organization.
```

File: github/orgs_packages.go [hunk 35]
```diff
@@ -0,0 +1,141 @@
...
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-version-for-an-organization
+func (s *OrganizationsService) PackageGetVersion(ctx context.Context, org, packageType, packageName string, packageVersionID int64) (*PackageVersion, *Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v/versions/%v", org, packageType, packageName, packageVersionID)
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	var version *PackageVersion
```
Comment (gmlewis) [inline github/orgs_packages.go:108]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 36]
```diff
@@ -0,0 +1,141 @@
...
+	resp, err := s.client.Do(ctx, req, &version)
+	if err != nil {
+		return nil, resp, err
+	}
+	return version, resp, nil
+
+}
+
+// Delete a package version from an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:117]:
```suggestion
// Delete a package version from an organization.
```

File: github/orgs_packages.go [hunk 37]
```diff
@@ -0,0 +1,141 @@
...
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-package-version-for-an-organization
+func (s *OrganizationsService) PackageDeleteVersion(ctx context.Context, org, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v/versions/%v", org, packageType, packageName, packageVersionID)
+	req, err := s.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/orgs_packages.go:126]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 38]
```diff
@@ -0,0 +1,141 @@
...
+
```
Comment (gmlewis) [inline github/orgs_packages.go:127]: Please delete this blank line.

File: github/orgs_packages.go [hunk 39]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Restore a package version to an organization
```
Comment (gmlewis) [inline github/orgs_packages.go:130]:
```suggestion
// Restore a package version to an organization.
```

File: github/orgs_packages.go [hunk 40]
```diff
@@ -0,0 +1,141 @@
...
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-package-version-for-an-organization
+func (s *OrganizationsService) PackageRestoreVersion(ctx context.Context, org, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	u := fmt.Sprintf("orgs/%v/packages/%v/%v/versions/%v/restore", org, packageType, packageName, packageVersionID)
+	req, err := s.client.NewRequest("POST", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	return s.client.Do(ctx, req, nil)
```
Comment (gmlewis) [inline github/orgs_packages.go:139]: Please add a blank line above this line for consistency with the rest of the repo.

File: github/orgs_packages.go [hunk 41]
```diff
@@ -0,0 +1,141 @@
...
+
```
Comment (gmlewis) [inline github/orgs_packages.go:140]: Please delete this blank line.

File: github/packages.go [hunk 42]
```diff
@@ -99,3 +106,40 @@ type PackageRegistry struct {
...
+	// Visibility of packages public, internal or private
+	Visibility string `url:"visibility,omitempty"`
+
+	// Type of package
+	PackageType string `url:"package_type,omitempty"`
+
+	// State of package either active or deleted
+	State string `url:"state,omitempty"`
+
+	// For paginated result sets, page of results to retrieve.
```
Comment (gmlewis) [inline github/packages.go:121]: Please replace lines 121-125 with the single line:

```
ListOptions
```

File: github/packages.go [hunk 43]
```diff
@@ -99,3 +106,40 @@ type PackageRegistry struct {
 func (r PackageRegistry) String() string {
 	return Stringify(r)
 }
+
+// PackageListOptions represents the optional list options for a package
```
Comment (gmlewis) [inline github/packages.go:110]:
```suggestion
// PackageListOptions represents the optional list options for a package.
```

File: github/packages.go [hunk 44]
```diff
@@ -99,3 +106,40 @@ type PackageRegistry struct {
...
+type PackageListOptions struct {
+	// Visibility of packages public, internal or private
+	Visibility string `url:"visibility,omitempty"`
+
+	// Type of package
+	PackageType string `url:"package_type,omitempty"`
```
Comment (gmlewis) [inline github/packages.go:116]:
```suggestion
	PackageType *string `url:"package_type,omitempty"`
```

File: github/packages.go [hunk 45]
```diff
@@ -99,3 +106,40 @@ type PackageRegistry struct {
...
+	State string `url:"state,omitempty"`
+
+	// For paginated result sets, page of results to retrieve.
+	Page int `url:"page,omitempty"`
+
+	// For paginated result sets, the number of results to include per page.
+	PerPage int `url:"per_page,omitempty"`
+}
+
+// PackageMetadata represents metadata from a package
```
Comment (gmlewis) [inline github/packages.go:128]:
```suggestion
// PackageMetadata represents metadata from a package.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @Cyb3r-Jak3 !
This is off to a great start.  I have to run now, but will continue my review later.
This should give you a good idea of what I'm looking for.

**Comment by gmlewis** [comment]
Thank you, @Cyb3r-Jak3 , but I'm going to hold off on a complete review until all unit tests are passing.
It looks like they are currently failing here:

```
    orgs_packages_test.go:208: Request parameters: map[state:[deleted]], want map[page:[2] per_page:[10] state:[deleted]]
--- FAIL: TestOrganizationsService_ListPackagesVersions (0.00s)
```

Please see our CONTRIBUTING.md guide to show you how to run the tests before pushing changes.  Thanks.

**Comment by Cyb3r-Jak3** [comment]
@gmlewis, sorry about that. All good now.

File: github/users_packages.go [hunk 46]
```diff
@@ -0,0 +1,207 @@
...
+// List the packages for a user. Passing the empty string for "user" will
+// list packages for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#list-packages-for-the-authenticated-users-namespace
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#list-packages-for-a-user
+func (s *UsersService) ListPackages(ctx context.Context, user string, opts *PackageListOptions) ([]*Package, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("user/%v/packages", user)
```
Comment (gmlewis) [inline github/users_packages.go:21]: Can you please add a unit test for the case when `user` is non-nil?  (`"u"` is fine, and common in other unit tests.)

File: github/users_packages.go [hunk 47]
```diff
@@ -0,0 +1,207 @@
...
+
+// Delete a package from a user. Passing the empty string for "user" will
+// delete the package for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-for-a-user
+func (s *UsersService) DeletePackage(ctx context.Context, user, packageType, packageName string) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v", user, packageType, packageName)
```
Comment (gmlewis) [inline github/users_packages.go:79]: Can you please add a unit test for the case when user is non-nil?

File: github/users_packages.go [hunk 48]
```diff
@@ -0,0 +1,207 @@
...
+
+// Restore a package to a user. Passing the empty string for "user" will
+// restore the package for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-for-a-user
+func (s *UsersService) RestorePackage(ctx context.Context, user, packageType, packageName string) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/restore", user, packageType, packageName)
```
Comment (gmlewis) [inline github/users_packages.go:100]: Can you please add a unit test for the case when user is non-nil?

File: github/users_packages.go [hunk 49]
```diff
@@ -0,0 +1,207 @@
...
+
+// Get all versions of a package for a user. Passing the empty string for "user" will
+// get versions for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-all-package-versions-for-a-package-owned-by-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/users#delete-an-email-address-for-the-authenticated-user
+func (s *UsersService) PackageGetAllVersions(ctx context.Context, user, packageType, packageName string) ([]*PackageVersion, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions", user, packageType, packageName)
```
Comment (gmlewis) [inline github/users_packages.go:121]: Can you please add a unit test for the case when user is non-nil?

File: github/users_packages.go [hunk 50]
```diff
@@ -0,0 +1,207 @@
...
+
+// Get a specific version of a package for a user. Passing the empty string for "user" will
+// get the version for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#get-a-package-version-for-a-user
+func (s *UsersService) PackageGetVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*PackageVersion, *Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v", user, packageType, packageName, packageVersionID)
```
Comment (gmlewis) [inline github/users_packages.go:148]: Can you please add a unit test for the case when user is non-nil?

File: github/users_packages.go [hunk 51]
```diff
@@ -0,0 +1,207 @@
...
+
+// Delete a package version for a user. Passing the empty string for "user" will
+// delete the version for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#delete-package-version-for-a-user
+func (s *UsersService) PackageDeleteVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v", user, packageType, packageName, packageVersionID)
```
Comment (gmlewis) [inline github/users_packages.go:175]: Can you please add a unit test for the case when user is non-nil?

File: github/users_packages.go [hunk 52]
```diff
@@ -0,0 +1,207 @@
...
+
+// Restore a package version to a user. Passing the empty string for "user" will
+// restore the version for the authenticated user.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-a-package-version-for-the-authenticated-user
+// GitHub API docs: https://docs.github.com/en/rest/reference/packages#restore-package-version-for-a-user
+func (s *UsersService) PackageRestoreVersion(ctx context.Context, user, packageType, packageName string, packageVersionID int64) (*Response, error) {
+	var u string
+	if user != "" {
+		u = fmt.Sprintf("users/%v/packages/%v/%v/versions/%v/restore", user, packageType, packageName, packageVersionID)
```
Comment (gmlewis) [inline github/users_packages.go:196]: Can you please add a unit test for the case when user is non-nil?

File: github/orgs_packages_test.go [hunk 53]
```diff
@@ -0,0 +1,368 @@
...
+			Type:              String("Organization"),
+			SiteAdmin:         Bool(false),
+		},
+	}}
+	if !cmp.Equal(packages, want) {
+		t.Errorf("Organizations.ListPackages returned %+v, want %+v", packages, want)
+	}
+
+	const methodName = "ListPackages"
+	testNewRequestAndDoFailure(t, methodName, client, func() (*Response, error) {
```
Comment (gmlewis) [inline github/orgs_packages_test.go:99]: For every unit test that you added, can you please follow the pattern of other tests in this repo and also call 

```
testBadOptions(t, methodName, func() (err error)
```

?  (Please see `repos_test.go` for many examples.)

This will help with the code coverage.  Thank you!

**Review by gmlewis — CHANGES_REQUESTED** [review]

File: github/users_packages_test.go [hunk 54]
```diff
@@ -52,10 +52,14 @@ func TestUsersService_ListPackages(t *testing.T) {
...
 	if !cmp.Equal(packages, want) {
-		t.Errorf("Users.ListPackages returned %+v, want %+v", packages, want)
+		t.Errorf("Users.Authenticated_ListPackages returned %+v, want %+v", packages, want)
 	}
 
 	const methodName = "ListPackages"
+	testBadOptions(t, methodName, func() (err error) {
+		_, _, err = client.Users.ListPackages(ctx, "\n", nil)
+		return err
+	})
```
Comment (gmlewis) [inline github/users_packages_test.go:62]:
```suggestion
	})

```

File: github/orgs_packages_test.go [hunk 55]
```diff
@@ -0,0 +1,408 @@
...
+			"url": "https://api.github.com/orgs/github/packages/container/hello_docker",
+			"created_at": `+referenceTimeStr+`,
+			"updated_at": `+referenceTimeStr+`,
+			"html_url": "https://github.com/orgs/github/packages/container/package/hello_docker"
+		  }
+		  ]`)
+	})
+
+	ctx := context.Background()
+	packages, _, err := client.Organizations.ListPackages(ctx, "o", nil)
```
Comment (gmlewis) [inline github/orgs_packages_test.go:58]:
```suggestion
	packages, _, err := client.Organizations.ListPackages(ctx, "o", &PackageListOptions{})
```

File: github/orgs_packages_test.go [hunk 56]
```diff
@@ -55,7 +55,7 @@ func TestOrganizationsService_ListPackages(t *testing.T) {
 	})
 
 	ctx := context.Background()
-	packages, _, err := client.Organizations.ListPackages(ctx, "o", nil)
+	packages, _, err := client.Organizations.ListPackages(ctx, "o", &PackageListOptions{})
```
Comment (gmlewis) [inline github/orgs_packages_test.go:58]: @Cyb3r-Jak3 - can you please update all the other tests to follow this pattern?

**Review by gmlewis — CHANGES_REQUESTED** [review]

**Review by gmlewis — APPROVED** [review]
Thank you, @Cyb3r-Jak3 !
This looks great!  I really appreciate your patience with me and my reviews.
LGTM.

Awaiting second LGTM before merging.

**Comment by Cyb3r-Jak3** [comment]
@wesleimp Nudge

**Comment by gmlewis** [comment]
@Parker77 - if you have time to do a code review, that would be greatly appreciated.

**Review by Parker77 — APPROVED** [review]
LGTM.

**Comment by gmlewis** [comment]
> LGTM.

Thank you, @Parker77 !
Merging.
